### PR TITLE
Refactor prerequisite checks in Developer CLI

### PR DESCRIPTION
**Pull request by PavlovicDzFilip**
### Summary & Motivation

Update the prerequisites configuration to be strongly typed in the Developer CLI, reducing boilerplate code and making the code more concise.

### Checklist

- [x] I have added a Label to the pull-request
- [x] I have added tests, and done manual regression tests
- [x] I have updated the documentation, if necessary

### CODE REVIEW

**Review by tjementum — CHANGES_REQUESTED** [review]
Nice one. Thanks. I made a few small comments.

Did you have any problems… or did you just see an opportunity for improvements?

The description mentions problems on Windows, but I don’t see anything specific to Windows? I’ve updated the title and PR description (I avoid using the term “we” and “this PR” when describing the changes as it goes into commit history).

**Comment by tjementum** [comment]
BTW... Please force push your changes. No need for one more commit ;)

File: developer-cli/Installation/PrerequisitesChecker.cs [hunk 1]
```diff
@@ -5,71 +5,36 @@
...
-public record Prerequisite(PrerequisiteType Type, string Name, string DisplayName, Version Version);
-
-public enum PrerequisiteType
-{
-    CommandLineTool,
-    DotnetWorkload
-}
-
-public static class PrerequisitesChecker
+public abstract record Prerequisite
```
Comment (tjementum) [inline developer-cli/Installation/PrerequisitesChecker.cs:8]: Please rename the file from `PrerequisitesChecker` to just `Prerequisite`

File: developer-cli/Installation/PrerequisitesChecker.cs [hunk 2]
```diff
@@ -5,71 +5,36 @@
...
+        var isValid = prerequisites.Aggregate(true, (state, prerequisite) => prerequisite.IsValid() && state);
+        if (!isValid)
         {
             Environment.Exit(1);
         }
     }
+}
 
-    private static bool IsCommandLineToolValid(string command, string displayName, Version minVersion)
+file record CommandLineToolPrerequisite(string Command, string DisplayName, Version MinVersion) : Prerequisite
```
Comment (tjementum) [inline developer-cli/Installation/PrerequisitesChecker.cs:29]: Could you make these two classes `sealed`. I get a warning in SonarLint.

CleanShot 2024-09-08 at 16 31 36

File: developer-cli/Commands/DevCommand.cs [hunk 3]
```diff
@@ -16,7 +16,12 @@ public DevCommand() : base("dev", "Run the Aspire AppHost with all self-containe
 
     private void Execute()
     {
-        PrerequisitesChecker.Check("dotnet", "docker", "aspire", "node");
+        Prerequisite.Ensure(
+            Prerequisite.Dotnet,
+            Prerequisite.Docker,
+            Prerequisite.Aspire,
+            Prerequisite.Node
+        );
```
[the remark this answers is not in the document]
Comment (PavlovicDzFilip) [reply developer-cli/Commands/DevCommand.cs]: Are there any other rules? And even better, is there a shared editor config to apply the rules automatically?

File: developer-cli/Installation/PrerequisitesChecker.cs [hunk 4]
```diff
@@ -5,71 +5,36 @@
...
-                    break;
-            }
-        }
+    protected abstract bool IsValid();
 
-        if (invalid)
+    public static void Ensure(params Prerequisite[] prerequisites)
+    {
+        var isValid = prerequisites.Aggregate(true, (state, prerequisite) => prerequisite.IsValid() && state);
+        if (!isValid)
```
[the remark this answers is not in the document]
Comment (PavlovicDzFilip) [reply developer-cli/Installation/PrerequisitesChecker.cs]: Having a variable with proper name is more readable, so I'd go with that.

**Comment by PavlovicDzFilip** [comment]
> Nice one. Thanks. I made a few small comments.
> 
> Did you have any problems… or did you just see an opportunity for improvements?
> 
> The description mentions problems on Windows, but I don’t see anything specific to Windows? I’ve updated the title and PR description (I avoid using the term “we” and “this PR” when describing the changes as it goes into commit history).

It does not run on Windows correctly. So far, I've found these issues: 
- [Checking if Docker is running does not work](https://github.com/platformplatform/PlatformPlatform/blob/main/developer-cli/Commands/DevCommand.cs#L23). The name of the process on Windows is `Docker Desktop`
- Command `open` is used to start some programs, or a web page - this command does not exist on Windows. 
  - [To start Docker](https://github.com/platformplatform/PlatformPlatform/blob/main/developer-cli/Commands/DevCommand.cs#L26), we need to pass in the location of the executable. The location would usually be `C:\Program Files\Docker\Docker\Docker Desktop.exe` which will be incorrectly split by [ProcessHelper.CreateProcessStartInfo](https://github.com/platformplatform/PlatformPlatform/blob/main/developer-cli/Utilities/ProcessHelper.cs#L51) method
  
  
I have not stumbled upon other issues so far to run the project. However, when running the Aspire Host, I was not able to debug the system - the breakpoints did not attach. Any idea why?

Finally: is there any community Slack/Discord channel to join?

File: developer-cli/Commands/DevCommand.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (tjementum) [reply developer-cli/Commands/DevCommand.cs]: There are many rules :) JetBrains rules (see `.dotsettings`) and SonarCloud (all default rules) are used for static code analysis.

You can run these commands:

- `pp code-inspections`
- `pp code-cleanup`
- `npm run format && npm run build && npm run check`

Or you can use R# or Rider to do a full solution code cleanup. I recommend “Format on save”. For the frontend, Biome is used over Prettier - it’s much faster and flexible. It’s a recommended plugin (see `.vscode` settings).

But any findings in these will break the build.

As for line breaks, originally we had a hard limit of 120 characters and used JetBrains rules. But everyone, including me, found that we did not always like the line breaks. So now this 120-length for any meaning full code is by convention, and manually enforced.

**Comment by tjementum** [comment]
OK. I have a few companies building on PlatformPlatform who are using Windows. But I believe the CLI is not used that much. I have, however, seen it work several times, so it’s weird that it does not behave consistently.

> Finally: is there any community Slack/Discord channel to join?
Could you try: https://join.slack.com/t/platformplatform/signup

PS: Feel free to convert this to "Ready for review". 

Also... did you not get a request to sign the CLA? I had some problems, but please do it here: https://cla-assistant.io/platformplatform/PlatformPlatform

**Comment by tjementum** [comment]
> Are there any other rules

I just merged your other pull request. We always rebase to get a nice Git commit graph, so please rebase and force push, and then I think it’s ready to be merged.

PS: I noticed that you signed the CLA in the closed pull request from yesterday. So, all good.

**Review by tjementum — APPROVED** [review]

**Comment by PavlovicDzFilip** [comment]
> > Could you try: https://join.slack.com/t/platformplatform/signup

I am being asked for an @platformplatform.net email address, which I do not have.

**Comment by tjementum** [comment]
> I am being asked for an @platformplatform.net email address, which I do not have.

Please try again. I just removed it (forgot to click save before).

**Comment by PavlovicDzFilip** [comment]
> Please try again. I just removed it (forgot to click save before).

It is still the case. 
image

**Comment by tjementum** [comment]
Ahh... it's because Slack is adding @platformplatform.net back. I should change a setting. Try one last time.

Sorry for the inconvenience.

**Comment by PavlovicDzFilip** [comment]
Now I need an invite: 

My email address is pavlovic.dz.filip@outlook.com (to be redacted)
image

**Comment by tjementum** [comment]
OK. It's because this feature is not supported by Slack.

But I created an invite link and wrapped in a nice looking url here: https://bit.ly/platformplatform-slack-invite.

Sorry I should have tested before I shared the links :)

**Pull request merged**